### ISSUE TITLE
lit-html + redux?

### ISSUE THREAD

**Issue by jmas** [PolymerLabs/lit-html#185]
Hello, I found that `lit-html` is built and using mostly in Polymer ecosystem, but I did few concepts to demonstrate that `lit-html` can be used with react-like libraries, for example `redux`.
Also I did a quick-port of famous library `Formik` that provide high-ordered component that help manage forms.

Here is few concepts:
* [lit-redux](https://github.com/jmas/lit-redux)
* [lit-form](https://github.com/jmas/lit-form)
* [lit-todo-example](https://github.com/jmas/lit-todo-example)

What you think about that use case of `lit-html`?

Also I collided with few questions:
* How to get access to DOM elements like we do in react-like components (often by `ref`)? 
* How to control lifecycle of DOM nodes? (It's useful when we try inject some plugins that use DOM nodes directly).

Thanks!

**Comment by NeoLegends** [comment]
We guys over at @Festify came up with something similar, based on mixins: https://github.com/Festify/fit-html.

**Comment by jmas** [comment]
@NeoLegends I impressed of work that was done for `fit-html` (p.s. I live TS, and envy that `fit-html` implemented with it).
Btw I tried to clone `fit-html-demo` and build it, but found that npm package `fit-html` is absent. It will be great to publish it, to make demo working.
Thanks for response!

**Comment by NeoLegends** [comment]
Ah yeah, the demo is built with local fit-html at the moment. Just clone fit-html in a directory beside the demo and compile the sources using „yarn build“. It’s all very rough (its a product of a mini-hackathon) at the moment, but we intend to publish fit-html soon. :)

**Comment by NeoLegends** [comment]
Respectively, while hacking the demo together, we just did a „yarn install ../fit-html“, which obviously doesn‘t really do a reproducible install, since you have to ensure the directory / the project is present.

**Comment by jmas** [comment]
@NeoLegends Thanks, it's a live! Found strange error `$$observable is not defined` but fixed that error by define globally this variable with value `1`. Will check your sources, maybe found interesting approaches or answers on my questions. Btw do you have some information about my questions?

**Comment by NeoLegends** [comment]
Yeah, you need to use a Browser that has symbol support, I think. We don't Babel anything in the demo right now, so I advise you use an evergreen browser with support for symbols to view the demo.

**Comment by NeoLegends** [comment]
> How to get access to DOM elements like we do in react-like components (often by ref)?

In `💪-html` you can always subclass the class you get from calling `connect`, which gives you access to all lifecycle methods and the rendering callbacks. You would override one of the methods and hook your DOM-node logic up in there. 

> How to control lifecycle of DOM nodes? (It's useful when we try inject some plugins that use DOM nodes directly).

As `💪-html` gives you a subclass of `HTMLElement`, you can just subclass that thing again and override whichever DOM lifecycle callbacks you want (`connectedCallback`, `disconnectedCallback`, etc.).

**Comment by NeoLegends** [comment]
@jmas `💪-html` is now on npm 🎉

**Comment by jmas** [comment]
@NeoLegends Oh, I found that directive function called with `part` that contain links to nodes range. I really far from Polymer, so I do not got basic concepts. I get that Custom Elements have two callbacks as you told before (`connectedCallback()`, `disconnectedCallback()`) but (for me) more intuitive to see the callbacks as directive options and it is not clear for me how to use it.

**Comment by NeoLegends** [comment]
> Oh, I found that directive function called with part that contain links to nodes range. 

I'm sorry but could you rephrase that? I don't get what you mean by that.

> I really far from Polymer, so I do not got basic concepts. I get that Custom Elements have two callbacks as you told before (connectedCallback(), disconnectedCallback()) but (for me) more intuitive to see the callbacks as directive options and it is not clear for me how to use it in current form.

You mean it would be more natural for you to pass the callbacks into the `connect`-function from the outside and have those called?

I don't think this is the best approach, as its quite limiting actually. Subclassing is superior IMO:
- You have more freedom. You can choose when and how your code and the code in the superclass runs.
- It is more aligned to the "use the platform" principle.
- The library is lighter because we have less logic to build.

**Comment by jmas** [comment]
@NeoLegends I did following test:
```
${ directive(part => { console.log(part); return html` ${ 'hello' } ` }) }
```
And get:
```
NodePart {
  instance: TemplateInstance,
  startNode: text,
  endNode: text,
  _previousValue: TemplateInstance
}
```

`startNode` and `endNode` point to first and last node of part wrapped by directive.

> You mean it would be more natural for you to pass the callbacks into the connect-function from the outside and have those called?

I mean to do something like that:
```
directive(part => { /* ... */ }, {
  connect(part) {},
  disconnect(part) {}
})
```

In this case I can control specific DOM elements that wrapped by directive without necessity to create Custom Element to use it `connectedCallback()`, `disconnectedCallback()` callbacks.

**Comment by NeoLegends** [comment]
I think this is out of the scope of lit-html. lit-html is meant to be a DOM-rendering library with template strings, not a component framework. If you want that, use custom elements or react.

**Comment by jmas** [comment]
@NeoLegends Ok, thanks for helping with that. I should learn more about Custom Elements then. In some meaning `react` is DOM-rendering library too. Maybe I get a false impression that `lit-html` = `react`, so I tried fit it to react-like environment.

**Comment by richardanaya** [comment]
While we are tossing around personal libraries ;) I borrowed the great ideas from recompose and lit-html and put them into a project WebCompose.

https://github.com/richardanaya/webcompose

**Comment by MatxBerg** [comment]
@jmas I was wondering what was triggering the "re-render" of the lit-html template after a modification is made in the todo app?. From what I understand, lit-html must re render every time a property is changed? I'm I wrong?

**Comment by obedm503** [comment]
it's worth mentioning [skatejs](https://github.com/skatejs/skatejs) in this. 
if using skate with the LirRenderer: lit-html does the rendering, skate provides the component model with lifecycle methods and stuff, and redux handles state management.

micro libraries ftw!
